vector/src/lib.rs

```diff
@@ -0,0 +1,30 @@
+//! This crate provides persistent data structures tailored to Nickel's needs.
+//!
+//! [`Vector`] is a persistent vector (also known as a "bitmapped vector trie")
+//! with cheap clones and efficient copy-on-write modifications. [`Slice`]
+//! backs the implementation of arrays in Nickel. It's basically a [`Vector`]
+//! with support for slicing.
```

Comment:
Mostly just https://hypirion.com/musings/understanding-persistent-vector-pt-1, which I linked a little later. I looked at `rpds` to see what choices they were making, but I didn't really imitate their implementation otherwise.